Add async debug-repl support

SevereOverfl0w opened the pull request:
About 2.5 years ago you wrote this for me, and told me to clean it up
and PR.  I finally got round to it!  I renamed it to async to indicate
it's general utility.

gfredericks commented on src/com/gfredericks/debug_repl/async.clj:
```diff
@@ -0,0 +1,59 @@
+  ;; if the repl is busy executing a request, should concurrent
+  ;; requests block or skip the repl execution altogether?
+  ;;
+  ;; going to try blocking first
```
this comment is pretty important -- probably best to make this function able to do both, based on an arg

SevereOverfl0w replied:
Doesn't this need to be controlled from `break!`, not from `wait-for-breaks`? Or am I misunderstanding the meaning here?

gfredericks replied:
I think you could do it either way, but I'm not sure how you would do it exactly; I think if you want to make it not-blocking, replacing the ArrayBlockingQueue with a SynchronousQueue would let you detect whether the `loop` can execute something; that might constrain it to be implemented in `wait-for-breaks`

SevereOverfl0w commented:
I've addressed most of the feedback, these commits can be squashed on merge as they're feedback ones, but separation helps keep the review history clearer on GitHub.

I'd like to know more about the skip vs block feature.  Right now the comment claims to block but I'd describe the code as skipping, so I must be misunderstanding something.

gfredericks commented on src/com/gfredericks/debug_repl/async.clj:
```diff
@@ -0,0 +1,65 @@
+    (resolve 'clojure.tools.nrepl.middleware.interruptible-eval/*msg*)))
+
+(defn can-break?
+  "Implementation detail. Subject to change."
```
why not just make it actually `^:private`?

SevereOverfl0w replied:
Because it's used from the macro below

gfredericks replied:
I guess `msg-var` can't be private either :/

SevereOverfl0w replied:
It can, because it's called by a public function, not by the macro directly.

Macros can't generate code with direct references to private vars (I assume because they're not available in the call-site).

gfredericks replied:
augh okay right, I'm just reading sloppily, sorry; I was imagining that `can-break?` could be inlined, at which point `msg-var` couldn't be private; so I guess it's one or the other

SevereOverfl0w commented:
Coming back to this block/unblocked with fresh eyes, I have better understanding.  The `(loop)`/`(recur)` is really important to implementing a lock-up in `wait-for-breaks`, and makes more sense to me now. Also I had previously assumed the executor was a a Java-thing, but now I see it's just a function, hence it controls how break! behaves.

The `.put` is the main place to decide whether the `break!` should wait or whether to continue in the background. If `(.offer)` was used instead, then it could time out, or not write if more than the 16 requests were pending.

I'm not certain if a ArrayBlockingQueue is needed here, the SynchronousQueue would always be used. You would then change whether the executor used `(.put)` to block & wait or `(.offer)` dependent on the behavior you want. You would rely on the already-threaded or sync behavior of the system to manage hand-offs between them. Especially as we essentially block until the work is complete anyway, now we would just block via `.offer` primarily, rather than `deref`.

---

I've understood this now, and pushed up an implementation.  The downside of the extra parameter is requiring the timeout to be specified.  Is it possible that we actually want 2 different functions here with different names?

gfredericks commented:
Okay, I think this looks good; thanks!
